src/rest/CommunityEndpoints.js

```diff
@@ -1,24 +1,200 @@
+
+    const key = typeof options.wallet === 'string' ? options.wallet : options.key || options.privateKey || client.options.auth.privateKey
+    if (key) {
+        const provider = options.provider instanceof providers.Provider ? options.provider : getDefaultProvider()
+        return new Wallet(key, provider)
+    }
+
+    // TODO: check metamask before erroring!
+    throw new Error("Please provide options.wallet, or options.privateKey string, if you're not authenticated using a privateKey")
+}
+
+// //////////////////////////////////////////////////////////////////
+//          admin: DEPLOY AND SETUP COMMUNITY
+// //////////////////////////////////////////////////////////////////
+
+/**
+ * Deploy a new CommunityProduct contract and create the required joinPartStream
+ * Note that the Promise resolves with an ethers.js TransactionResponse, so it's only sent to the chain at that point, but not yet deployed
+ * @param {EthereumOptions} options such as blockFreezePeriodSeconds (default: 0), adminFee (default: 0)
+ * @return {Promise<Contract>} has methods that can be awaited: contract is deployed (`.deployed()`), operator is started (`.isReady()`)
+ */
+export async function deployCommunity(options) {
+    const wallet = parseWalletFromOptions(this, options)
+
+    const blockFreezePeriodSeconds = options.blockFreezePeriodSeconds || 0
+    const adminFee = options.adminFee || 0
+    const tokenAddress = options.tokenAddress || this.options.tokenAddress
+    const streamrNodeAddress = options.streamrNodeAddress || this.options.streamrNodeAddress
+    const streamrOperatorAddress = options.streamrOperatorAddress || this.options.streamrOperatorAddress
+    // const {blockFreezePeriodSeconds, adminFeeFraction, tokenAddress, streamrNodeAddress, operatorAddress} = {
+    //     blockFreezePeriodSeconds: 0,
+    //     adminFee: 0,
+    //     ...this.options,
+    //     ...options,
+    // }
+
+    await throwIfNotContract(wallet.provider, tokenAddress, 'options.tokenAddress')
+    await throwIfBadAddress(streamrNodeAddress, 'options.streamrNodeAddress')
+    await throwIfBadAddress(streamrOperatorAddress, 'options.streamrOperatorAddress')
+
+    if (adminFee < 0 || adminFee > 1) { throw new Error('options.adminFeeFraction must be a number between 0...1, got: ' + adminFee) }
+    const adminFeeBN = new BigNumber((adminFee * 1e18).toFixed()) // last 2...3 decimals are going to be gibberish
+
+    const stream = await this.createStream({
+        name: `Join-Part-${wallet.address.slice(0, 10)}-${Date.now()}`
+    })
+    debug(`Stream created: ${JSON.stringify(stream.toObject())}`)
+    const res1 = await stream.grantPermission('read', null)
+    debug(`Grant read permission response from server: ${JSON.stringify(res1)}`)
+    const res2 = await stream.grantPermission('write', streamrNodeAddress)
+    debug(`Grant write permission response to ${streamrNodeAddress} from server: ${JSON.stringify(res2)}`)
```

Comment:
Could we maybe extract this to a separate function? It would help us to unify `joinPartStream` creation on frontend since we cannot use `deployCommunity` directly because we need to watch the transaction.

Reply:
yes, although this `deployCommunity` works a bit differently from the old one: this one returns immediately with the ethers.js Transaction object (you can get the tx hash and the future contract address), and you can then either `await contract.deployed()` or even `await contract.isReady()` which means server has noticed the creation and started operating, so it's ready to use.

Reply:
but you may be right about how it works together with UI stuff like Metamask popup. I imagine Metamask popup would appear at `deployer.deploy` call which creates the tx. But the function would return right after user gives the OK, and waiting for the tx would then be optional.

Reply:
the issue with separating those is that they still need to be done together both or none, so if separation is needed, then probably better just duplicate the code, almost... depends on the case for sure.

Reply:
We are using `web3` on the frontend side currently so that Ethers object might not help.

We have this code currently duplicated on the frontend side and it's not that bad. The name we use for the stream is a bit different on our side.